fuel_gauge: Add fuel_guage_get_prop/s()

aaronemassey opened the pull request:
The fuel_gauge_get_prop() function prototype declares a function that retrieves multiple fuel gauge properties at once. The naming suggests it ought to fetch a singular property at a time. Moreso, some clients may just want to fetch properties one at a time and may feel uncomfortable using a prototype for fetching multiple properties when wanting to fetch them one at a time.

Modify fuel_gauge_get_prop() to fetch a single property and add fuel_gauge_get_props() to support fetching multiple properties. Modify existing tests/drivers/samples.

This is part of #61818 work.

aaronemassey committed:
fuel_gauge: Add fuel_guage_get_props()

The fuel_gauge_get_prop() function prototype declares a function that
retrieves multiple fuel gauge properties at once. The naming suggests it
ought to fetch a singular property at a time. Moreso, some clients may just
want to fetch properties one at a time and may feel uncomfortable using a
prototype for fetching multiple properties when wanting to fetch them one
at a time.

Modify fuel_gauge_get_prop() to fetch a single property and add
fuel_gauge_get_props() to support fetching multiple properties. Modify
existing tests/drivers/samples.

This is part of #61818 work.

Signed-off-by: Aaron Massey <aaronmassey@google.com>

aaronemassey commented:
Doing this before https://github.com/zephyrproject-rtos/zephyr/pull/62366 - SBS Property Moving.

fabiobaltieri commented on drivers/fuel_gauge/bq27z746/bq27z746.c:
```diff
@@ -307,7 +291,7 @@ static int bq27z746_init(const struct device *dev)
 static const struct fuel_gauge_driver_api bq27z746_driver_api = {
-	.get_property = &bq27z746_get_props,
+	.get_property = &bq27z746_get_prop,
 	.set_property = &bq27z746_set_props,
```
doing get and set separately?

aaronemassey replied:
That was the plan. But I can combine them into a single PR if you'd like - normally I try to do 1 commit = 1 PR when reasonable. I should've called that out more explicitly in the top comment.

fabiobaltieri replied:
the change is probably symmetric, I think it make sense to do both at once, up to you tough

aaronemassey replied:
I'll do both at once then!

fabiobaltieri commented on include/zephyr/drivers/fuel_gauge.h:
```diff
@@ -255,6 +264,7 @@ typedef int (*fuel_gauge_battery_cutoff_t)(const struct device *dev);
 
 __subsystem struct fuel_gauge_driver_api {
 	fuel_gauge_get_property_t get_property;
+	fuel_gauge_get_properties_t get_properties;
```
I don't think this is a good idea, you'd end up driver implementing only one of the two and it'll all become a big mess, just have `get_properties` outside of the api struct and leave only the subsystem implementation. Let's clean up the drivers with a funny props implementation separately.

aaronemassey replied:
> you'd end up driver implementing only one of the two and it'll all become a big mess,

Is there something wrong with a driver only implementing one of the two? IMO, that's not terribly messy. How about I hide this behind a common config like: `CONFIG_FUEL_GAUGE_CUSTOM_GET_PROPS_FN` ?

fabiobaltieri replied:
yeah, one could implement the multi but not the single and then applications using the single would not work with it, makes no sense, but also like keeping it simple, there's no need for the dual api

aaronemassey replied:
As discussed offline: Adding comment to explicitly call out the multi-get/set API field as a possible future direction but leaving it out until we have a concrete upstream use-case.

fabiobaltieri commented on include/zephyr/drivers/fuel_gauge.h:
```diff
@@ -264,29 +274,64 @@ __subsystem struct fuel_gauge_driver_api {
+	int err_count = 0;
+	const struct fuel_gauge_driver_api *api = (const struct fuel_gauge_driver_api *)dev->api;
+
+	/* Allow for driver optimized versions. */
+	if (api->get_properties) {
+		return api->get_properties(dev, props, len);
+	}
```
see my comment above, I don't think this is a good idea

aaronemassey committed:
fuel_gauge: Add fuel_guage_set_props()

The fuel_gauge_set_prop() function prototype declares a function that sets
multiple fuel gauge properties at once. The naming suggests it ought to
fetch a singular property at a time. Moreso, some clients may just want to
set properties one at a time and may feel uncomfortable using a prototype
for fetching multiple properties when wanting to fetch them one at a time.

Modify fuel_gauge_set_prop() to fetch a single property and add
fuel_gauge_set_props() to support fetching multiple properties. Modify
existing tests/drivers/samples.

This is part of #61818 work.

Signed-off-by: Aaron Massey <aaronmassey@google.com>

fabiobaltieri commented on include/zephyr/drivers/fuel_gauge.h:
```diff
@@ -222,7 +222,7 @@ struct sbs_gauge_device_chemistry {
  */
 typedef int (*fuel_gauge_get_property_t)(const struct device *dev,
-					 struct fuel_gauge_property *props, size_t props_len);
+					 struct fuel_gauge_property *prop);
```
Cool but I think this now should take `(dev, prop_type, &prop_value)` rather than the composite structure with the status bit that this does not need anyway now that it has a return code. This way getting a property become much more compact as you can just put the type in the call itself and use an uninitialized value as argument.

aaronemassey replied:
Oh I was going to do that in a follow-up PR after removing the exit code. Want me to separate out the property type out in in this PR?

fabiobaltieri replied:
your choice but I think that our downstreaming person would appreciate to get the api changes in a single pr

aaronemassey replied:
Then I'll do it in the same PR, along with the exit code. I'll keep things up as separate commits so we can change them to separate PRs easily if wanted.

aaronemassey committed:
fuel_gauge: Remove status from fuel gauge properties

Based on review of the similar charger driver API, it's been demonstrated
from the community that embedding a per value status code when fetching
multiple properties isn't particularly wanted or needed. It was largely
considered not worth the additional maintenance to have the extra per
property error information.

Remove the status field from the fuel_gauge property value structs.

Signed-off-by: Aaron Massey <aaronmassey@google.com>

aaronemassey committed:
fuel_gauge: Repl property struct w/ union

Based on review of the similar charger driver API, it's been demonstrated
from the community that embedding a per value property type when fetching
properties. Separating off the property types from the property values
themselves also allow an array of property types to declared as static
const.

Break up fuel_gauge_property struct into a fuel_gauge_prop_val union and a
fuel_gauge_prop_t property type as inputs into fuel gauge API functions.

Signed-off-by: Aaron Massey <aaronmassey@google.com>

fabiobaltieri commented on drivers/fuel_gauge/max17048/max17048.c:
```diff
@@ -175,23 +175,24 @@ static int max17048_init(const struct device *dev)
  * Get a single property from the fuel gauge
  */
-static int max17048_get_single_prop_impl(const struct device *dev, struct fuel_gauge_property *prop)
+static int max17048_get_single_prop_impl(const struct device *dev, fuel_gauge_prop_t property_type,
+					 union fuel_gauge_prop_val *prop)
```
Maybe use `prop` and `val` for `fuel_gauge_prop_t` and `union fuel_gauge_prop_val`?

fabiobaltieri commented on include/zephyr/drivers/fuel_gauge.h:
```diff
@@ -215,17 +204,17 @@ struct sbs_gauge_device_chemistry {
  */
-typedef int (*fuel_gauge_get_property_t)(const struct device *dev,
-					 struct fuel_gauge_property *prop);
+typedef int (*fuel_gauge_get_property_t)(const struct device *dev, fuel_gauge_prop_t prop_type,
+					 union fuel_gauge_prop_val *prop);
```
I'd standardize all of these to `prop` and `val`, same as charger, sensors has like `chan, val` or `attr, val`...

fabiobaltieri reviewed (commented):
Thanks, just a nitpick on the naming

teburd commented on drivers/fuel_gauge/bq27z746/bq27z746.c:
```diff
@@ -109,101 +109,100 @@ static int bq27z746_read_mac(const struct device *dev, uint16_t cmd, uint8_t *da
-		rc = bq27z746_read16(dev, BQ27Z746_AVERAGECURRENT, &val);
-		prop->value.avg_current = (int16_t)val * 1000;
+		rc = bq27z746_read16(dev, BQ27Z746_AVERAGECURRENT, &tmp_val);
+		val->avg_current = (int16_t)tmp_val * 1000;
```
This is nice

teburd reviewed (dismissed):
This looks good to me, and simplifies some things seemingly which is nice

keith-zephyr commented on doc/hardware/peripherals/fuel_gauge.rst:
```diff
@@ -20,8 +20,9 @@ Fundamentally, a property is a quantity that a fuel gauge device can measure.
 
-Properties are fetched using a client allocated array of :c:struct:`fuel_gauge_property`.  This
-array is then populated by values as according to its `property_type` field.
+Properties are fetched one at a time using a client allocated :c:struct:`fuel_gauge_property` or
+fetched all at once using a client allocated array of :c:struct:`fuel_gauge_property`. Each
+:c:struct:`fuel_gauge_property` is populated by values according to its `property_type` field.
```
I find it clearer to tell the user the function call to make.

```suggestion
Properties are fetched by the client one at a time using :c:func:`fuel_gauge_get_prop`, or fetched in a batch using :c:func:`fuel_gauge_get_props`.  These functions fill in the :c:struct:`fuel_gauge_property` allocated by the client according to its `property_type` field.
```

aaronemassey replied:
Whoops. Missed this.

aaronemassey replied:
I corrected the documentation based on the final changes in the last commit. Is that acceptable?

keith-zephyr commented on drivers/fuel_gauge/bq27z746/bq27z746.c:
Nit - change commit message title to "fuel_gauge: Remove status from fuel gauge properties"

keith-zephyr commented on tests/drivers/fuel_gauge/bq27z746/src/test_bq27z746.c:
```diff
@@ -42,10 +42,7 @@ ZTEST_USER_F(bq27z746, test_get_all_props_failed_returns_negative)
-		      props[0].property_type);
-
-	zassert_true(ret < 0);
+	zassert_equal(ret, -ENOTSUP);
```
Note - I think `-EINVAL` is probably a more appropriate error code, but I see that the existing drivers use `-ENOTSUP`.

aaronemassey replied:
I think it's not entirely clear which one to use here. As you mentioned, sensor version uses `-ENOTSUP` and there may be properties that are entirely valid but just aren't supported by this driver.
